repository_service_tuf_worker/repository.py

```diff
@@ -607,108 +679,388 @@ def save_settings(self, root: Metadata[Root], roles_info: Dict[str, Any]):
+            raise RepositoryError(
+                "Delegations already using hash-bins, cannot add custom roles"
             )
-            # Initialize all succinct delegated roles (`bins`), update expire,
-            # sign, add to `Snapshot` meta and persist in the backend storage
-            # service.
-            for delegated_name in succinct_roles.get_roles():
-                targets.signed.add_key(online_pub_key, delegated_name)
-                bins_role = Metadata(Targets())
-                self._bump_expiry(bins_role, BINS)
-                self._sign(bins_role)
-                self._persist(bins_role, delegated_name)
-                delegated_roles.append(delegated_name)
-
-        self.write_repository_settings(
-            "DELEGATED_ROLES_NAMES", delegated_roles
-        )
+
+        success = {}
+        failed = []
+        db_roles = []
+
+        logging.debug("adding roles to Targets delegations")
+        for role in delegations.roles:
+            if targets.signed.delegations.roles is None:
+                logging.debug("initializing Target delegations")
+                targets.signed.delegations.roles = {}
+
+            if (
+                role in targets.signed.delegations.roles
+                or targets_crud.read_role_deactivated_by_rolename(
+                    self._db, role
+                )
+            ):
+                logging.info(f"Role '{role}' already exists, skipping")
+                failed.append(
+                    {
+                        "role": role,
+                        "reason": (
+                            "role already exists or name used in the past",
+                        ),
+                    }
+                )
+                continue
+
+            # create delegated target role
+            role_metadata = Metadata(Targets(version=1))
+
+            # save expiration settings
+            expires = delegations.roles[role].unrecognized_fields[
+                "x-rstuf-expire-policy"
+            ]
+            self.write_repository_settings(
+                f"{role.upper()}_EXPIRATION", expires
+            )
+
+            # add keys to the delegated target role
+            # if no key is assigned, use the online key
+            if (
+                len(delegations.roles[role].keyids) == 0
+                and delegations.roles[role].threshold > 1
+            ):
+                failed.append(
+                    {
+                        "role": role,
+                        "reason": "If no keys assigned threshold must be 1",
+                    }
+                )
+
+            self._add_delegated_role_keys(
+                targets, delegations, role, role_metadata
+            )
+
+            if role not in targets.signed.delegations.roles:
+                logging.info(f"Role '{role}' added to Targets delegations")
+                targets.signed.delegations.roles[role] = delegations.roles[
+                    role
+                ]
```

Comment:
If you call `add_metadata_delegations` then when will it happen that this `role` will be part from `targets.signed.delegations.roles`? 
Seems like this `if` will be always true.

Reply:
When the user tries to add an existing role.

Also, if you see the `targets.signed.delegations.roles` is a list of DelegatedRoles, the user "could" add the same role twice.